src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+                        Sse.StoreScalar(pDstCurrent, src2Vector);
+
+                        pSrc1Current++;
+                        pSrc2Current++;
+                        pDstCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static float SumSqU(float[] src, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        result = Sse.Add(result, Sse.Multiply(srcVector, srcVector));
+
+                        pSrcCurrent += 4;
+                    }
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
+                    while (pSrcCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                        result = Sse.AddScalar(result, Sse.MultiplyScalar(srcVector, srcVector));
+
+                        pSrcCurrent++;
+                    }
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static float SumAbsU(float[] src, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+            Vector128<float> mask;
+
+            if (Sse2.IsSupported)
+            {
+                mask = Sse.StaticCast<int, float>(Sse2.SetAllVector128(0x7FFFFFFF));
+            }
+            else
+            {
+                mask = Sse.SetAllVector128(0x7FFFFFFF);
+            }
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        result = Sse.Add(result, Sse.And(srcVector, mask));
+
+                        pSrcCurrent += 4;
+                    }
+
+                    // TODO: Check code length per line
+                    if (Sse3.IsSupported)
```

Comment:
Abstracting this out into a helper `VectorSum` method might be useful (it is used here and above)

Reply:
Thanks - fixed with a helper `VectorSum` method with return type `Vector128<float>`.  Do you think it's better for `VectorSum` to change the input array in place and have the return type be `void` instead?

Reply:
No, returning `Vector128<float>` is better here as the existing code you pulled out into the helper method doesn't modify the array.